controller/controller.go

```diff
@@ -238,6 +262,37 @@ func (c *Controller) RunOnce(ctx context.Context) error {
 	return nil
 }
 
+func splitChanges(changes *plan.Changes) (*plan.Changes, *plan.Changes, *plan.Changes) {
```

Comment:
I think it might become a little easier (especially the splitting code) if we directly work with the endpoint lists, i.e. changing the signature to:

```go
func splitChanges(changes *plan.Changes) ([]endpoint.Endpoint, []endpoint.Endpoint, []endpoint.Endpoint)
```

Reply:
If we work directly on the slices instead of the Changes struct, then the `applyHalfActionChanges` code will become something like this:
```
func (p *Provider) ApplyChanges(changes) {
  err := applyChanges(changes) # make calls to AWS etc.
  if err != nil {
     # split into the actions (low hanging fruit to reduce the batch. not sure if it's always correct, though)
     create, update, delete := separateActions(changes)
     
     # retry only the record creations
     applyChanges(Changes{Create: create})
     
     # similarly for updates and deletes (note, update are actually two lists)
     # applyChanges(Changes{Update: update})
     # applyChanges(changes{Delete: delete})

     # splitting would just split the []endpoint.Endpoint slice in half and retry each side.
     left, right := splitInHalf(create)
     applyChanges(changes{left})
     applyChanges(changes{right})     

     # keep splitting the half further etc. (this logic should probably be done differently, e.g. in the next sync iteration)
     left2, right2 := splitInHalf(left)
     applyChanges(changes{left2})
     applyChanges(changes{right2}) 
     
     ....
  }
}